linked_list.py

```diff
@@ -0,0 +1,348 @@
+
+        if not(node) or node.get_data() > value:
+            self.insert(value)
+            return
+
+        while node.next_node and node.next_node.get_data() < value:
+            node = node.next_node
+
+        new_node = Node(value = value, next_node = node.next_node)
+        node.next_node = new_node
+
+
+    # method to print all the values in the linked list
+    def visit(self):
+        current_node = self.__head
+        while True:
+            print(current_node.get_data())
+            if current_node.next_node == None: break
+            current_node = current_node.next_node
+
+    # # method to delete the first node found with specified value
+    def delete(self, value):
+        node = self.__head
+        if not(node): return
+        if node.get_data() == value:
+            self.__head = node.next_node
+
+        while node.next_node:
+            if node.next_node.get_data() == value:
+                node.next_node = node.next_node.next_node
+                break
+            node = node.next_node
+
+
+    # # method to reverse the singly linked list
+    # # note: the nodes should be moved and not just the values in the nodes
+    def reverse(self):
+        if self.__head == None or self.__head.next_node == None:
+            return
+
+        node = self.__head
+        prev_node = None
+
+        while node != None:
+
+            # preserve next node for later use
+            next_node = node.next_node
+
+            # make node point to the value before it
+            node.next_node = prev_node
+
+            # move forward one node
+            prev_node = node
+            node = next_node
+        self.__head = prev_node
+        self.visit()
+
+
+    # ## Advanced Exercises
+    # # returns the value at the middle element in the singly linked list
+    def find_middle_value(self):
+        node = self.__head
+        if not node:
+            return
+        else:
+            mid_val = node.get_data()
+
+        while node.next_node and node.next_node.next_node:
+            mid_val = node.next_node.get_data()
+            node = node.next_node.next_node
+
+        return mid_val
+
+    # # find the nth node from the  and return its value
+    # # assume indexing starts at 0 while counting to n
+    def find_nth_from_beginning(self, n):
+        if self.__head == None:
+            return "Undefined"
+        node = self.__head
+
+        for i in range(n):
+            if node.next_node:
+                node = node.next_node
+            else:
+                return "None"
+        return node.get_data()
+
+    # # checks if the linked list has a cycle. A cycle exists if any node in the
+    # # linked list links to a node already visited.
+    # # returns true if a cycle is found, false otherwise.
+    def has_cycle(self):
+        if not self.__head:
+            return False
+        slow_loop = self.__head
+        fast_loop = self.__head.next_node
+        while fast_loop.next_node.next_node:

```

Comment:
What happens if fast_loop.next_node is None:? null dereference issue. Check for fast_loop.next_node before checking for fast_loop.next_node.next_node.